Add a web browser dose viewer gui to EGSnrc

ellieb opened the pull request:
Add the web browser-based .egsphant, .3ddose, and DICOM file viewer. This project is a WIP, but the idea is to have an online resource for users to upload and view their own .egsphant and .3ddose files. There are many TODOs throughout the code that need to be addressed, but some of the major tasks include:

- [ ] Organize and restructure the code to be cleaner and more inline with web standards
- [ ] Add tests
- [ ] Do a visual cleanup of the site, making it more user friendly
- [X] Allow for multiple volume viewers to be functional at once
- [ ] Turn the window and slider bars into a double range slider on the density legend
- [ ] Fix the bug where the panel marker moves if changing a slice after zoom

ellieb committed:
Initial commit for web-based dose viewer

ellieb committed:
Add readers for .egsphant and .3ddose files

ellieb committed:
Add class to create a volume viewer

ellieb committed:
Add usage instructions to README

ellieb committed:
Add css file to style the dose viewer

ellieb committed:
Add support for view plane change

ellieb committed:
Add slice number data slider and Cartesian axes

Add a slider selector in the gui to changed the displayed phantom slice.
Fix phantom processing to ensure data outputs properly and scale data to
fit the canvas. Add Cartesian axes.

ellieb committed:
Update .3ddose file reader

ellieb committed:
Add dose contours

ellieb committed:
Add error array and store dose as sparse array

ellieb committed:
Clean up get slice and only redraw axes if needed

ellieb committed:
Create volume base class

ellieb committed:
Add legend for dose and density levels

ellieb committed:
Adjust appearance of dose contours and density map

Fill dose contours, update the colour map and fix the aliased appearance
of the density map. Also, fix a runtime issue due to a port closed error
left unchecked.

ellieb committed:
Add dose profile plotting

ellieb committed:
Add voxel coordinates and marker upon mouse click

Add voxel coordinates info upon mouse click, and a corresponding marker.
Display both world and voxel coordinates, as well as dose and density
information. Also, add an isEmpty function to the volume class.

ellieb committed:
Add increment and decrement buttons to slider

ellieb committed:
Add zooming capabilities to the dose viewer gui

Allow for zooming in and out of the voxel grid and displayed dose
contours. Ensure that the marker and voxel info behave as expected upon
zooming, and that the current zoom persists upon slice and axis change.

ellieb committed:
Improve the density and dose plot display

Add a density plot checkbox to the dose profile, change the colour
scales and set the plot background to black. Calculate contour
thresholds if the density is continuous. Add full-size crosshairs for
the marker, and update the slice number slider continuously.

ellieb committed:
Improve the dose profile plot code and display

Add a density option, the error range, and dose labels. Also, add an
object for dose profile dimensions and a getDoseProfileData function.

ellieb committed:
Add dose profiles to crosshairs

ellieb committed:
Add dose uncertainty to voxel information

ellieb committed:
Create a dose profile object

ellieb committed:
Extract zoom logic from index.html into zoom.js

ellieb committed:
Add zooming capabilities for dose profiles

ellieb committed:
Add the ability to export view to a png image

ellieb committed:
Fix issues with density and dose plots

Use get functions for slice and axis. Fix marker shifting after slice
change, and axis scaling issue. Fix density legend overflow.

ellieb committed:
Add enable/disable checkbox for dose profile plots

ellieb committed:
Convert size variables into objects

ellieb committed:
Maintain image aspect ratio and add axes grid

ellieb committed:
Improve the presentation of dose plots

Show plot axes if dose is uploaded first, identify the contour lines as
a percentage of the maximum dose, and allow for longer dose legends by
adding a scroll (set to auto).

ellieb committed:
Improve the layout of dose profile plots

Improve the layout of dose profile plots, add a grid, and change scales
of the profiles to percentage of the maximum dose.

ellieb committed:
Allow selecting interval between dose contours

ellieb committed:
Improve zooming in dose profile plots

Fix a bug in dose profile zooming, and use generic zooming functions

ellieb committed:
Use world coordinates in dose profile plot title

ellieb committed:
Define a dose profile data variable to return data

ellieb committed:
Improve the style of dose plots

Change the dose uncertainty in the information box to a percentage, add
axis labels on the main viewer (without the Gy unit).

ellieb committed:
Use a d3 scaling object to scale contours

ellieb committed:
Improve style and behavior of dose plot contours

Show and hide dose contours, persist hiding contours when changing
slices or adding additional lines, and add user-defined contours. Remove
dose contour interval selection and remove dose legend title on redraw.
Add max dose slider connected to dose plot, as percentage, with maximum
set to 150%. Fix incorrect voxel coordinates upon slider change. Clean
up axes, add ticks and adjust label position. Flip main viewer y-axis
and y pixel range. Make contour lines thinner. Also, use a gradient for
the density legend.

ellieb committed:
Improve style and behavior of dose profile plots

Add dimension unit "cm" to dose profile coordinates, and update dose
profiles when the maximum dose changes. Disable zoom when the dose
profile checkbox is unchecked and always replot when it is checked.

ellieb committed:
Export images as png and dose profile data as csv

Export images in png format instead of svg; hide the marker to export as
png, and disable the png export button until data is loaded. Allow
exporting dose profile data as a list of comma-separated values (csv).
Also, stack buttons and checkboxes to clean up display.

ellieb committed:
Enable file drag and drop to load data

Data can now be loaded by dragging-and-dropping a file on the web
application. Also, a file loading progress bar is added, and it is
hidden after loading is complete.

ellieb committed:
Add material to voxel data

ellieb committed:
Use an html canvas for the density and dose plot

Use an html canvas to display the voxel density map and dose data.
Adjust the zoom functionality to work with the canvas, and change the
variable name from svgDensity to canvDensity.

ellieb committed:
Add density level and window data sliders

ellieb committed:
Improve voxel information box and dose positioning

Fix the voxel information that was offset by one, and the position of
the dose data. Also, improve the style of the voxel information box, and
add a "show marker" checkbox to show or hide it.

ellieb committed:
Display the 3 plots for x, y and z axes at once

Show the density and dose data in three panels, one for each axis, all
at once, instead of a single axis view at a time. Display all three axes
on load, but only initialize the dose contour legend once.

ellieb committed:
Improve the style of dose and profile plots

Fix the dose and density legends, don't interpolate dose contours, and
use dose percentage in the voxel information box. Also, hide the dose
profiles plots at the bottom by default,

ellieb committed:
Add panel class and allow dragging of crosshairs

ellieb committed:
Add zooming to panel, after object initialization

ellieb committed:
Show correct voxel coordinates on drag

ellieb committed:
Add axis to prevSlice and prevSliceImg

ellieb committed:
Move through slices by clicking on the panel

ellieb committed:
Improve panel style and fix panel zoom

Add active marker class, fix axes scale on zoom and increase zoom extent

ellieb committed:
Sync cursor on click and marker position on drop

ellieb committed:
Improve layout of density and profile plots

Update the layout, add side scrolling for each volume, add a z axis to
the dose profile plots, and preserve the order of the profiles.

ellieb committed:
Fix dose profile zoom and drag marker behavior

Fix zooming for the dose profiles, and the drag marker behavior in the
zoomed panel. Also, fix the show/hide checkbox for dose profiles and the
voxel information.

ellieb committed:
Add a slider class, and a slider for each panel

Use a slider class. Add sliders for each panel, update sliders of other
panels on click, and update voxel information and marker position on
slider change.

ellieb committed:
Define a volume viewer object

ellieb committed:
Adapt dose viewer elements to volume viewer object

ellieb committed:
Fix slice number sliders and voxel information

ellieb committed:
Support dose profiles within volume viewer object

ellieb committed:
Fix zoom, and marker drag and drop behavior

ellieb committed:
Improve layout of voxel info and dose profiles

Update voxel information display and build its html using d3. Move file
selectors and voxel information inside of volume holder div, and move
dose profiles to bottom of panel.

ellieb committed:
Add the ability to load multiple volume viewers

Add a button to load multiple viewers on the web page, hide dose profile
if checkbox unchecked, and verify that the dose profile is loaded into
the volume viewer before plotting.

ellieb committed:
Load voxel and dose data at correct slice and zoom

ellieb committed:
Add dose file comparison dropdown selection

ellieb committed:
Enable comparisons between dose distributions

Enable the ability to load and compare multiple dose distributions.
Ensure that the dose contour input box and the max dose slider are not
duplicated, and maintain the zoom level on max dose change or dose
contour addition. Remove loaded volumes if deselected.

ellieb committed:
Clean up code and layout of the dose viewer

Use slider class for window and level sliders. Remove world-to-voxel
scales, prevAxis property and global variable densityVol. Disable zoom
until volume is loaded, and remove legacy code for dose profile with
custom inputs. Clean up voxel coordinates style, balance the look of
dose and density legends and hide the plot density checkbox. Allow line
breaks in dose files. Fix constructor argument ordering issue.

ellieb committed:
Export with dom-to-image instead of html2canvas

ellieb committed:
Load on center slice and open new viewer if needed

Set the initial slice position in center of volume, even if the first
volume is dose data. Also, open a new volume viewer for the first dose
or density volume loaded.

ellieb committed:
Add the option to load test file data

ellieb committed:
Add University of Victoria XCITE lab logo

ellieb committed:
Rename create-volume.js script file to volume.js

ellieb committed:
Comment and document the dose viewer code

Add comments in the javascript files to document the code for clarity.
Also, rename the checkbox hints, update the arguments of the voxel
information function, use relative paths for test files, and add
acknowledgments in readme file.

ellieb committed:
Clean up and polish dose viewer gui elements

Hide button to add new volume (for now), align slice number sliders with
main viewer, add margins to clean up display, center align slice number
sliders, and fix level and window sliders.

ellieb committed:
Add EGSnrc headers and update web page title

ftessier commented:
Rebased @ellieb's branch on the EGSnrc repo develop branch.

rtownson commented:
@ftessier should we instead be adding this code to the `gh-pages` branch? If I want to link from `gh-pages` to the dose viewer I will have to move a copy of the code there anyways, so there would be duplication.

ftessier commented:
Nope, we want the code to be in the repository for source control; We don't want to duplicate source management, pull requests, etc.,  on another branch; even though as you said implies that we duplicate code to the `gh-pages`. We can however create a new branch off of `gh-pages`, were we can cherry-pick the dose-viewer commits as they are added, so as to not have to manually copy files over everytime. If you put the code on `gh-pages`, for the moment I would leave it in a "hidden" folder (not linked on the web page), but accessible by url, until we test it a little...

ftessier commented:
@rtownson I take that back, I see now that we _could_ maintain the web dose viewer code on the orphan `gh-pages` branch, and create the pull requests against that branch witout issue. However, I still prefer to leave it attached to the main trunk, I feel like it belongs there as part of the code proper, and if the dose viewer development ever requires changes to other parts of the code (say, the DOSXYZnrc output format), then it becomes tricky to synchronize the two disjointed branches. I am sure there are services to host a webpage from a github repo directory; I will look into it!

ftessier commented:
Ok, well it seems that hosting with github pages remains the best option here. We can also now host a separate repo for `egsnrc.github.io` (instead of a separate _branch_), which may prove useful in terms of filtering out old pdfs (or maybe _not_ filtering them out!), and building a more complex site hosting the viewer etc. We could then create pull requests against that repo from the EGSnrc one; anyways, let's continue to think about this.

ftessier commented:
@ellieb I am considering using JS `standard` (https://standardjs.com/) to lint and format the `.js` files in your pull request
to format the code consistently, beyond personnal coding style. This avoids style divergence and meaningless diffs in the long run, especially as other people with different preferences start changing the code. We do this on C and C++ files with `astyle` for example. JS Standard seems to be the simplest option out there right now: once installed via `npm`, just issue `standard` to check all `.js` files in the directory. 

I ran it in the `dose-viewer` directory, and although there are 2340 suggestions, the vast majority are simply `Strings must use singlequote.` and `Extra semicolon.` "errors", which can be fixed automatically by `standard`. I just want to check that you are fine with this?

ftessier commented:
@ellieb There are only 41 remaining linting instances that seem to require some attention... At any rate, I will look into those after cleaning and squashing the branch, an consult with you at that point on the remaining issues noted by `standard`.

ftessier commented:
And I suggest we use `prettier` to lint html code (https://prettier.io/). Other suggestions welcome!

ftessier committed:
Format dose viewer code with Standard and Prettier

ftessier commented:
@ellieb I have reduced this pull request to 78 commits, squashing together some smaller, related commits. I think the resulting chain of commits makes logical sense. Please take a look when you have some time, and let me know if this is ok with you. I also adjusted the commit titles accordingly, but left most of the original titles in the commit messages body. The last commit `Format...` is meant to be amended to format the `.js` files with `standard` and `.html` files with `Prettier`.

ftessier commented:
@ellieb can you review the following log from the JS `standard` linter, and let me know if any one of these suggested fixes should _**not**_ be implemented? [standard.log](https://github.com/nrc-cnrc/EGSnrc/files/5214308/standard.log)

ftessier committed:
Format the javascript files with StandardJS

Run the standard command with the --fix option (see standardjs.com) to
check, lint and format the javascript files according to the StandardJS
style, automatically. After code modifications, run `standard` again to
curb style changes that are not relevant to code functionality.

ftessier committed:
Manually adjust the format of javascript files

Adjust the javascript code to curb StandardJS warnings: add global
definitions at the top of the files, add some esling-disable-line
comments for "defined but not used errors, change some "==" operators to
"===", and change a few variable names to camel case. The standard
command now runs cleanly in the dose viewer directory.

ftessier committed:
Adjust html and css style with Prettier formatter

ellieb committed:
Use ES6 import and export statements

ellieb committed:
View single DICOM file

ellieb committed:
Remove old window and level sliders on volume change

ellieb committed:
Add min density value to volume

ellieb committed:
Change window level density sliders to min max density sliders

ellieb committed:
Remove unnecessary console log statement

ellieb committed:
Read in multiple DICOM files

ellieb committed:
Change density formatting depending on if DICOM or egsphant file

ellieb committed:
Flip the y axis on the xy panel

ellieb committed:
Lock min and max sliders so they cannot pass eachother

ellieb committed:
Support DICOM dose files

ellieb committed:
Handle multi-frame dose DICOM files

ellieb committed:
Correctly display dose and density with different dimensions and move drawAxes to volumeViewer

ellieb committed:
Add min density to test file data

ellieb committed:
Fix bug in error of uneven offset length in pixel data

ellieb committed:
Remove extra console logging on DICOM file upload

ellieb committed:
Do not plot density error if it does not exist

ellieb committed:
Change updateVoxelCoords to use positional coordinates rather than voxel coordinatets

ellieb committed:
Fix bug in reading egsphant files

ellieb committed:
Add slice and density image caching

ellieb committed:
Automatically cache density slices

ellieb committed:
Use array to create density images to solve resolution issue

ellieb committed:
Add web worker to cache density slices in background

ellieb committed:
Allow slice caching in different browsers

ellieb committed:
Fix min and max sliders to work with cached slices

ellieb committed:
Fix slice number navigation bug

ellieb committed:
Update density legend upon min ma density change

ellieb committed:
Load both density and dose files upon upload

ellieb committed:
Allow for removal of dose and density files from volume

ellieb committed:
Replace invertTransform function with d3 invert function

ellieb committed:
Add baseSlices object to minimize slice calculations and remove prevSlice object from volume to improve behaviour with multiple volume viewers

ellieb committed:
Clean up cache worker

ellieb committed:
Move dose and density options inside the volume viewer

ellieb committed:
Redraw slice after changing dose and density sliders

ellieb committed:
Change density legend ticks instead of colour bar

ellieb committed:
Adjust dose scalings if corresponding density file has different voxel size

ellieb committed:
Always display dose file if one or more is uploaded

ellieb committed:
Add encoding to html file

ellieb committed:
Move from module code to script code

ellieb committed:
Don't make new volume viewer when loading test files

ellieb committed:
Add global functions to file upload

ellieb committed:
Fix bug where adding dose percentage does not show up

ellieb committed:
Filter out unrecognized file extensions instead of rejecting all files

ellieb committed:
Pass in svg to draw and clear dose and density

ellieb committed:
Fix zoom for multiple volume viewers

ellieb committed:
Add global imports and ignore unused vars to be inline with js standard

ellieb committed:
Fix show voxel and plot dose profile checkboxes to work with multiple volume viewers

ellieb committed:
Let min and max density sliders to work with multiple volume viewers

ellieb committed:
Let max dose slider work with multiple volume viewers

ellieb committed:
Let max dose in volume viewer work with dose profile plots

ellieb committed:
Allow export buttons to work with multiple volume viewers

ellieb committed:
Update DoseComparisonVolume and add the minDose attribute to the volume viewer and dose profile objects

ellieb committed:
Only load comparable doses in dose comparison dropdown

ellieb committed:
Adjust dose profile y axis scale for dose and dose comparisons

ellieb committed:
Fix bug in coordinates shifting on dose volume change

ellieb committed:
Calculate dose comparisons between RT Dose and 3ddose files

ellieb committed:
Fix bug in dose error display

ellieb committed:
Fix dose comparison selector to show all avaliable doses to compare

ellieb committed:
Fix density legend colourbar bug

ellieb committed:
Adjust opacity of dose contours

ellieb committed:
Implement trilinear interpolation method

ellieb committed:
Move code for clearing dose and density plots from volume to panel

ellieb committed:
Move initialize canvas logic to panel

ellieb committed:
Move density legend code into volume viewer

ellieb committed:
Read in RT Structure Set DICOM files

ellieb committed:
Add StructureSetVolume object

ellieb committed:
Get basic ROI contours working

ellieb committed:
Add ROI legend to volume viewer

ellieb committed:
Add zoom for ROI contours

ellieb committed:
Add function to build colour legends

ellieb committed:
Index slice sliders starting at 1

ellieb committed:
Adjust increment of ROI array to fix gap

ellieb committed:
Update ROI contours checkbox text

ellieb committed:
Add function to get volume data at world position

ellieb committed:
Add DVH calculation function to structure set volume object

ellieb committed:
Fix bug where mutiple polygons in one slice are not drawn

ellieb committed:
Increase width of ROI legend

ellieb committed:
Do not include unnecessary information in structure set volume

ftessier commented:
I am pushing that to Release 2022, and I suggest that we include it as a submodule.

ellieb committed:
Use scanline algorithm to generate ROI array to speed up generation

ellieb committed:
Bug fix for polygon scanline algo

ellieb committed:
Fix ROI display bug by indexing by given ROI number

ellieb committed:
Don't delete ROIs when adding dose contours

ellieb committed:
Fix bug with dose profile not showing for dose comparisons

ellieb committed:
Add DoseVolumeHistogram class

ellieb committed:
Add functionality for DVHs

ellieb committed:
Move cumulative histogram calculations to the setting data method in the DVH class

ellieb committed:
Add export DVH to CSV button

ellieb committed:
Return absolute dose from get data function

ellieb committed:
Convert DVH from relative to absolute dose

ellieb committed:
Don't delete ROI contours when adjusting dose contours

ellieb committed:
Use generic enable and disable checkbox functions

ellieb committed:
Add normalize dose checkbox and input boxes

ellieb committed:
Add dose normalization support for DVHs

ellieb committed:
Disable and uncheck DVH checkbox when ROI contours is unselected

ellieb committed:
Only show DVH checkbox if dose volume loaded and show ROI contours checkbox is selected

ellieb committed:
Add zoom for DVH plot

ellieb committed:
Include StudyInstanceUID in volume information

ellieb committed:
Match stucture set to displayed density/dose

ellieb committed:
Properly convert from Gy to cGy for dose normalization

ellieb committed:
Add dose units to dose volumes

ellieb committed:
Update checkbox behaviours to better work together

ellieb committed:
Simplify zoom for dose profiles

ellieb committed:
Minor css style changes

ellieb committed:
Add dose comparison normalization slider

ellieb committed:
Remove max dose and dose comparison normalization sliders when not useable

ellieb committed:
Add function to simply update voxel information from volume viewer

ellieb committed:
Organize volumes to simplify file structure

ellieb committed:
Organize sliders to simplify file structure

ellieb committed:
Fix spelling and casing errors

ellieb committed:
Update test files

ellieb committed:
Add slider file for slice slider

ellieb committed:
Fix RT DICOM scaling issue

ellieb committed:
Adjust font size

ellieb committed:
Only keep scroll for ROI legend

ellieb committed:
Update colours used for plot grids and labels

ellieb committed:
Line up lengend ends

ellieb committed:
Export dose profiles in single CSV file

ellieb committed:
Update min and max density slider labels

ellieb committed:
Change cursor to pointer over panels

mainegra commented:
@rtownson and @ftessier I am going through PRs and was wondering if this one is to be merged despite the WIP label?
